ggml/src/ggml-vulkan.cpp

```diff
@@ -1053,11 +1057,11 @@ static vk_buffer ggml_vk_create_buffer(vk_device& device, size_t size, vk::Memor
 
     uint32_t memory_type_index = UINT32_MAX;
 
-    memory_type_index = find_properties(&mem_props, &mem_req, req_flags);
+    memory_type_index = find_properties(&mem_props, &mem_req, req_flags, device->force_heap_index);
     buf->memory_property_flags = req_flags;
 
     if (memory_type_index == UINT32_MAX && fallback_flags) {
-        memory_type_index = find_properties(&mem_props, &mem_req, fallback_flags);
+        memory_type_index = find_properties(&mem_props, &mem_req, fallback_flags, device->force_heap_index);
         buf->memory_property_flags = fallback_flags;
     }
```

Comment:
Shouldn't this fallback already handle that situation? That's basically what it was introduced for. You have a smaller (DEVICE_LOCAL) and a larger heap (DEVICE_LOCAL, HOST_VISIBLE and HOST_COHERENT), so once the small one is exhausted you fall back to the larger one.
Can you add the vulkaninfo of the device where this fallback does not work and describe why?

Reply:
[vulkaninfo.txt](https://github.com/user-attachments/files/17263776/vulkaninfo.txt)

The problem is that `find_properties` doesn't take in account about memory usage (as `vk::PhysicalDeviceMemoryProperties` reports total mem, not available mem)
On boot, with a few chrome tabs open, my laptop consumes 2.8/4GiB of the smaller heap, so if it tries to allocate 2GiB, it will fail.

One could argue the better fix is to either
- use `VK_EXT_memory_budget` for querying actual memory usage in `find_properties`, so it doesn't select a memory type does not have enough available memory
- use a loop while allocating, if the allocation fails with `ErrorOutOfDeviceMemory`, try the next available memory type
- or a combination of both

I did the minimal viable solution here, which is to force allocation from a known bigger heap (in my case, heap 1).

Reply:
@gyf304 Thank you.

Take a look at https://github.com/ggerganov/whisper.cpp/pull/2451. This might be the kind of fallback that you described in your second point. Can you check whether that would fix your issue?

Reply:
FYI, I just pulled https://github.com/ggerganov/whisper.cpp/pull/2451 and will soon sync it upstream with `ggml` and `llama.cpp`. Not sure if it fixes the problem here, but it worked for a couple of `whisper.cpp`-related issues.